pkg/peertls/peertls.go

```diff
@@ -117,21 +123,45 @@ func VerifyPeerCertChains(_ [][]byte, parsedChains [][]*x509.Certificate) error
 	return verifyChainSignatures(parsedChains[0])
 }
 
-// VerifyCAWhitelist verifies that the peer identity's leaf was signed by any one of the
-// (certificate authority) certificates in the provided whitelist
-func VerifyCAWhitelist(cas []*x509.Certificate) PeerCertVerificationFunc {
+// VerifyCAWhitelist verifies that the peer identity's CA and leaf-extension was signed
+// by any one of the (certificate authority) certificates in the provided whitelist
+func VerifyCAWhitelist(cas []*x509.Certificate, verifyExtension bool) PeerCertVerificationFunc {
 	if cas == nil {
 		return nil
 	}
 
 	return func(_ [][]byte, parsedChains [][]*x509.Certificate) error {
-		var err error
+		var (
+			leaf = parsedChains[0][0]
+			err  error
+		)
+
+		// Leaf extension must contain leaf signature, signed by a CA in the whitelist.
+		// That *same* CA must also have signed the leaf's parent cert (regular cert chain signature, not extension).
 		for _, ca := range cas {
-			err = verifyCertSignature(ca, parsedChains[0][0])
+			err = verifyCertSignature(ca, parsedChains[0][1])
 			if err == nil {
```

Comment:
This might look nicer with `if err != nil { continue }` and `if !verifyExtension { break }`